add tutorial about SSLClientAuth with Apache 2.4 (#62)
hetzneronline/community-content: tutorials/ApacheSslClientAuth.md

## Step 1 - Install Apache web server and OpenSSL

The first step you need to do is to install the Apache web server and openssl (e. g. using "apt"). You only need to do this if you haven't installed the Apache web server and openssl before.
```bash
apt install apache2 openssl ssl-cert
```
If you're asked if you want to continue, please select "Y".

When this process is finished, please try to reach your IP address inside your web browser (e. g. http://10.0.0.1). You should now be able to see the default website of the Apache web server.

## Step 2 - Activate ssl module for apache2

The next step would be activating the ssl module for Apache. This module was already installed when running the installation of Apache as described in Step 1.
```bash
a2enmod ssl
```

If everything worked fine without any errors, you need to restart the Apache web server to load the modules that you have activated before.
```bash
systemctl restart apache2
```

You can check if the modules were loaded properly by running the following command:
```bash
apache2ctl -M
```
Please check if you can find the following lines inside the output:
```bash
[...]
ssl_module (shared)
[...]
```

> ## Step 3 - Get Let'sEncrypt certificate for basic SSL connections

Reviewer: ```suggestion
## Step 3 - Get Let's Encrypt certificate for basic SSL connections
```

> We now need to install the Let'sEncrypt certbot:

Reviewer: ```suggestion
We now need to install the Let's Encrypt Certbot:
```

```bash
apt-get install certbot python-certbot-apache
```

We'll now get the certificate:
```bash
certbot --apache
```

> You need to provide your mail address to get a certificate and you are asked to agree Let'sEncrypts Terms of Service. The Certbot tool also asks, if you would like to share your mail address with the developers of Certbot. It's up to you if you want to do that.

Reviewer: ```suggestion
You need to provide your mail address to get a certificate and you are asked to agree Let's Encrypts Terms of Service. The Certbot tool also asks, if you would like to share your mail address with the developers of Certbot. It's up to you if you want to do that.
```

When you run the above command, you will be asked for some data, such as your name and your location. It is recommended to fill this out correctly (but it is not necessary). Most important for SSL certificates is the common name, we recommend setting this to the domain name that you want to secure with this certificate.

## Step 5 - Generate client certificates

With the CA created above, you will now be able to generate client certificates. At first we need a key (such as in step 3 for the CA generation).
```bash
openssl genrsa -out MyClientCert1.key 4096
```

Now with that new key, we need to create a certificate signing request (CSR) - which is necessary for creating certificates using a CA.
```bash
openssl req -new -key MyClientCert1.key -out MyClientCert1.csr
```
* "req" stands for a new request to be made
* "-new" stands for generating a new CSR
* "-key MyClientCert1.key" stands for the key file that should be used - make sure to use the key that you have created for the client certificate and not the key for the CA certificate!
* "-out MyClientCert1.csr" specifies the location where the CSR should be saved

> When you run the above command, you will be asked for some data - such as before when creating the CA. The data that you provide here is more important here, especially the common name, since this can be used for authentication purposes later (so that not every certificate signed with your CA can access every web application that is secured with this CA). - We recommend using a username and your domain name as common name (e. g. holu@example.com). Additional attributes, such as "challenge password" and "optional company name" can be left empty.

Reviewer: ```suggestion
When you run the above command, you will be asked for some data - such as before when creating the CA. The data that you provide is more important here, especially the common name, since this can be used for authentication purposes later (so that not every certificate signed with your CA can access every web application which is secured with this CA). - We recommend using a username and your domain name as common name (e. g. holu@example.com). Additional attributes, such as "challenge password" and "optional company name" can be left empty.
```

Download the .pfx-file (e. g. via SCP) and provide it to the user, which has to use the certificate for login. The user should import this certificate into his web browser.

Please repeat the procedure of Step 4 for every client certificate that you want to create.

## Step 6 - Add rules for SSL authentictaion in your virtual host (or .htaccess)

Please add the following rules to your .htaccess file or your virtual host file to activate the SSL client authentication for several directories or complete websites. .htaccess files should always be located in the directory with the files the .htaccess rules should be applied to.

```
SSLVerifyClient require
```
This enables the SSL client authentication.

```
SSLVerifyDepth 10
```
This sets the depth of the verification. It needs to be set if the server needs to check more certificates. (This is the case if chained certificates are used.)

You may also limit the access to specific certificates using "SSLRequire". You can find more detailed information in the offical Apache 2 documentation (see: https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslrequire).

## Step 7 - Prepare directory for CA files

> Now, we need to create a new directory for the SSL-CAs and make it accessible for Apache to read from that directory.

Reviewer: ```suggestion
Now, we need to create a new directory for the SSL-CAs and make it possible for Apache to read from that directory.
```
